File: source/kl/render/Plane.java
package kl.render;

import kl.math.*;

public class Plane {
    public Float3 normal;
    public Float3 point;

    public Plane() {
        normal = new Float3();
        point = new Float3();
    }
    public Plane(Float3 normal, Float3 point) {
        this.normal = normal;
        this.point = point;
    }
    public Plane(Plane p) {
        normal = new Float3(p.normal);
        point = new Float3(p.point);
    }

    public String toString() {
        return "{" + normal + ", " + point + "}";
    }
}

File: klibrary/source/math/ray/plane.cpp
#include "klibrary.h"


// Class
kl::Plane::Plane()
{}

kl::Plane::Plane(const Float3& origin, const Float3& normal)
    : origin(origin)
{
    set_normal(normal);
}

// Normal
void kl::Plane::set_normal(const Float3& normal)
{
    m_normal = normalize(normal);
}

kl::Float3 kl::Plane::normal() const
{
    return m_normal;
}

// Math
bool kl::Plane::in_front(const Float3& point) const
{
    const float result = dot(point - origin, m_normal);
    return (result >= 0.0f);
}

// Format
std::ostream& kl::operator<<(std::ostream& os, const kl::Plane& obj)
{
    os << "{" << obj.origin << ", " << obj.normal() << "}";
    return os;
}
